src/daft-recordbatch/src/ops/window.rs

```diff
@@ -89,8 +89,139 @@ impl RecordBatch {
         self.union(&window_result)
     }
 
+    #[allow(clippy::too_many_arguments)]
+    pub fn window_agg_dynamic_frame(
+        &self,
+        name: String,
+        agg_expr: &AggExpr,
+        min_periods: i64,
+        dtype: &DataType,
+        frame: &WindowFrame,
+    ) -> DaftResult<Self> {
+        if matches!(frame.frame_type, WindowFrameType::Range) {
+            return Err(DaftError::ValueError(
+                "RANGE frame type is not supported yet for window_agg_dynamic_frame".into(),
+            ));
+        }
+
+        let total_rows = self.len();
+
+        // Determine if we have dynamic or static boundaries
+        let is_start_dynamic = matches!(frame.start, WindowBoundary::Offset(_));
+        let is_end_dynamic = matches!(frame.end, WindowBoundary::Offset(_));
+
+        // Calculate static bounds (if applicable)
+        let static_start = if !is_start_dynamic {
+            match &frame.start {
+                WindowBoundary::UnboundedPreceding() => Some(0),
+                WindowBoundary::UnboundedFollowing() => {
+                    return Err(DaftError::ValueError(
+                        "UNBOUNDED FOLLOWING is not valid as a starting frame boundary".into(),
+                    ));
+                }
+                _ => None,
+            }
+        } else {
+            None
+        };
+
+        let static_end = if !is_end_dynamic {
+            match &frame.end {
+                WindowBoundary::UnboundedFollowing() => Some(total_rows),
+                WindowBoundary::UnboundedPreceding() => {
+                    return Err(DaftError::ValueError(
+                        "UNBOUNDED PRECEDING is not valid as an ending frame boundary".into(),
+                    ));
+                }
+                _ => None,
+            }
+        } else {
+            None
+        };
+
+        self.window_agg_non_incremental(
+            agg_expr,
+            &name,
+            dtype,
+            frame,
+            static_start,
+            static_end,
+            min_periods,
+            total_rows,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn window_agg_non_incremental(
+        &self,
+        agg_expr: &AggExpr,
+        name: &str,
+        dtype: &DataType,
+        frame: &WindowFrame,
+        static_start: Option<usize>,
+        static_end: Option<usize>,
```

Comment:
In my mind it makes more sense if this function instead receives two options where None means either unbounded preceding/following and Some(x) means offset by x. And then, you do not need to pass in the frame, and can avoid the `match { .., _ => unreachable!(..) }` since you've restricted the types properly.